### ISSUE TITLE
MacOS: bad install names for shared libs (libasmrun_shared.so etc.)

### ISSUE THREAD

**Issue by mobileink** [ocaml/ocaml#12979]
On MacOS, shared libraries embed an "install name" as well as a list of shared library dependencies.  They can be inspected with `otool -L`.

In OPAM installations of the OCaml distribution, the install names are incorrect.  For example:

```
~ $ otool -L .opam/5.0.0/lib/ocaml/libasmrun_shared.so
.opam/5.0.0/lib/ocaml/libasmrun_shared.so:
	runtime/libasmrun_shared.so (compatibility version 0.0.0, current version 0.0.0) <1>
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1319.0.0)
```
<1> The install name is `runtime/libasmrun_shared.so`

Similarly for `libcamlrun_shared.so`.

This breaks shared libraries that link against these libraries.  This is because such libraries embed the install name.  For example, if we build shared library `dllalpha_stubs_bundled_lib_shared_rt.so` against `libcamlrun_shared.so` we get something like this:

```
$ otool -L ../stublibs/dllalpha_stubs_bundled_lib_shared_rt.so
../stublibs/dllalpha_stubs_bundled_lib_shared_rt.so:
	../stublibs/dllalpha_stubs_bundled_lib_shared_rt.so (compatibility version 0.0.0, current version 0.0.0)
	runtime/libcamlrun_shared.so (compatibility version 0.0.0, current version 0.0.0) <1>
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1336.61.1)
```
<1> install name of dependency treated as path

Then when we try to load `dllalpha_stubs_bundled_lib_shared_rt.so` we will get an error: _... Library not loaded: runtime/libcamlrun_shared.so...Reason: tried: 'runtime/libcamlrun_shared.so' (no such file)..._

I've created a suite of MWEs that you can use to verify this, at [https://github.com/obazl/demos_obazl/tree/main/makefiles/ffi]( https://github.com/obazl/demos_obazl/tree/main/makefiles/ffi).  Clone the repo, then `cd makefiles/ffi/dynamic`, then 

    make vm.unbundled.shared
    ./alpha.vm
    otool -L ../stublibs/dllalpha_stubs_bundled_lib_shared_rt.so

I'm not sure what the proper fix for this is.  A workaround that will only work during development is to change the install name to the absolute path of the library. This can be done with the `install_name_tool` utility; for example:

     ~ $ install_name_tool -id "$HOME/.opam/5.0.0/lib/ocaml/libcamlrun_shared.so" .opam/5.0.0/lib/ocaml/libcamlrun_shared.so

After that, running `.alpha.vm` as above should work.

**Comment by nojb** [comment]
Would it work to remove the directory part from the install name (ie have the install name be only `libasmrun_shared.so`)?

**Comment by damiendoligez** [comment]
>A workaround that will only work during development is to change the install name to the absolute path of the library. This can be done with the `install_name_tool` utility

Is this the return of the `ranlib` business under another guise?

It looks like the proper fix would be for the Makefile to call `install_name_tool` after installing the files.

**Comment by mobileink** [comment]
> Would it work to remove the directory part from the install name (ie have the install name be only `libasmrun_shared.so`)?

Nope, tried that.  I also fiddled around with `@executable_path`, `@loader_path`, and `@rpath` (see [Linking and Install Names](https://www.mikeash.com/pyblog/friday-qa-2009-11-06-linking-and-install-names.html)) but I don't think those will work (they seem to be designed for Frameworks).  It looks to me like an absolute path is the only thing that works.

**Comment by mobileink** [comment]
> It looks like the proper fix would be for the Makefile to call `install_name_tool` after installing the files.

That's what it looks like to me too.

The DSOs in `stublibs` might need the same treatment.

**Comment by tmcgilchrist** [comment]
@mobileink The instructions for running `make vm.unbundled.shared` seem to reference a make target that doesn't exist. Could you advise what the correct target is? I'm looking at the instructions for this version https://github.com/obazl/demos_obazl/blob/8733cc7dc716b8d5ae5a17199684ce2cbe8be6ef/makefiles/ffi/README.adoc

**Comment by tmcgilchrist** [comment]
Please keep this open. It is relevant to obazl and possibly future
debugging support.

**Comment by arbipher** [comment]
Is this issue still relevant? I have some experience in solving (macOS) linking problems and would like to help.
